Counterexamples As Assumptions

Dargones opened the pull request:
### Problem

Right now Dafny reports counterexamples using special syntax that may be difficult to understand. For example, consider the following Dafny program that defines a standard `List` datatype with a function `View` that maps the list to a sequence of integers:

```Dafny
datatype Node = Cons(next:Node, value:int) | Nil {
  function View():seq<int> 
    ensures Cons? ==> |View()| > 0 && View()[0] == value && View()[1..] == next.View()
  {
    if Nil? then [] else [value] + next.View()
  }
}
```

Suppose we were to (incorrectly) assert that the list cannot correspond to the sequence of integers `[1, 2, 3]`, like so:

```Dafny
method m(list:Node) {
  assert list.View() != [1, 2, 3];
}
```

Currently, Dafny would return the following counterexample: 

```
At "method m(list:Node) {" (file.dfy:20):
    list:Problem.Node = Cons(next := @0, value := 1)
    @0:Problem.Node = Cons(next := @2, value := 2)
    @2:Problem.Node = Cons(next := @4, value := 3)
    @4:Problem.Node = Nil
```

This counterexample is confusing because it does not explain what the meaning of `@0`, `@1`, etc. is. The notation is different from the Dafny syntax, and it also does not capture some of the information that is actually contained within the counterexample because there is no way to express this information using this custom syntax. In particular, the counterexample constrains the value returned by calling `View()` on the `list` variable. This constrain might be redundant in this particular example but in general, we would want to capture all the information contained in the counterexample.

### Solution

This PR redesigns the counterexample generation functionality so that counterexamples are represented internally and can be printed as Dafny assumptions. For example, for the program above, the counterexample will now look like this:

```Dafny
assume Node.Cons(Node.Cons(Node.Cons(Node.Nil, 3), 2), 1) == list 
    && Node.Cons(Node.Cons(Node.Cons(Node.Nil, 3), 2), 1).View.requires() 
    && [1, 2, 3] == Node.Cons(Node.Cons(Node.Cons(Node.Nil, 3), 2), 1).View() 
    && Node.Cons(Node.Cons(Node.Nil, 3), 2).View.requires() 
    && [2, 3] == Node.Cons(Node.Cons(Node.Nil, 3), 2).View() 
    && Node.Cons(Node.Nil, 3).View.requires() 
    && [3] == Node.Cons(Node.Nil, 3).View() 
    && Node.Nil.View.requires() 
    && [] == Node.Nil.View();
```

While admittedly more verbose because it explores the return values of functions, this counterexample precisely constrains the argument that leads to the assertion violation. In particular, if you were to insert this assumption into the code and revert the assertion, Dafny would verify that this counterexample is correct.

At a high level, this makes the following changes:
1) Represent counterexamples as Dafny statements that the user can insert directly into their code to debug the failing assertion.
2) Make sure the counterexamples are wellformed, i.e. the constraints are ordered in such a way that the resulting assumption verifies.
3) Support counterexample constraints over function return values (as an example of something that can really only be done using native Dafny syntax)
4) Automatically fix counterexamples that are internally inconsistent, when such an inconsistency can be easily detected. For instance, a counterexample will never describe negative indices of an array or sequence, call a set empty if it contains elements, etc. All of these are possible in the counterexample model returned by Boogie but we prune such inconsistencies before reporting the counterexample, when possible.


By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

atomb commented on Source/DafnyLanguageServer/CounterExampleGeneration/PartialState.cs:
```diff
@@ -0,0 +1,331 @@
+  }
+
+  /// <summary>
+  /// Convert this counterexample state into an assumption that could be inserted in Dafny source code
```
I like how simple this method is.

atomb commented on Source/DafnyLanguageServer/CounterExampleGeneration/DafnyModel.cs:
```diff
@@ -418,369 +754,105 @@ public class DafnyModel {
 
     /// <summary>
-    /// Return all functions mapping an object to a destructor value.
+    /// Return all functions application relevant to an element. 
```
```suggestion
    /// Return all function applications relevant to an element. 
```

atomb reviewed (changes requested):
This is an impressive bit of work! I had a few questions about it which could lead to changes, or that I'd like to at least know the answers to before merging it.

keyboardDrummer commented:
> The special syntax used by counterexamples is also not expressive enough to support all kinds of constraints that a Dafny counterexample might require

Can you elaborate? What kinds of constraints?

> This pull request addresses this issue by representing counterexamples as Dafny assumptions.

In addition to the existing method, or as a separate feature. Can you elaborate in the PR details what the functional changes of this PR are?

Dargones commented on Source/DafnyLanguageServer.Test/Various/CounterExampleTest.cs:
```diff
@@ -746,11 +725,11 @@ class Value {
     }
 
-    [Theory]
+    [Theory(Skip = "This test should be re-enabled once counterexamples support arrays")]
```
Have now added support for constraining arrays! They require some special processing since they are some array-specific AST nodes but it should all work now.

Dargones commented on Source/DafnyLanguageServer/CounterExampleGeneration/PartialState.cs:
```diff
@@ -0,0 +1,331 @@
+  /// variables that represent fields of any object in the original set or
+  /// elements of any sequence in the original set, etc. This is done
+  /// recursively in breadth-first order and only up to a certain maximum
+  /// depth.
```
I decided to get rid of the `depth` parameter altogether in my most recent commit. In practice, we would always want to call this method with `-1` (i.e. no depth limit), since we want all the information from the counterexample that we can get. I think the original motivation for this argument was to potentially limit the size of the counterexample displayed in the IDE. That doesn't actually help though since it only confuses the user by not displaying all of the information, and I think for that reason the Language Server does not use this argument any more (only the old server for Dafny 2 does).

Dargones commented:
Thank you, @atomb and @keyboardDrummer for the feedback on this! I have addressed the comments in my most recent commits and also updated the PR description to give a concrete example and a longer overview of the changes.

atomb commented on Source/IntegrationTests/TestFiles/LitTests/LitTest/server/counterexample_commandline.dfy.expect:
```diff
@@ -3,23 +3,13 @@ counterexample_commandline.dfy(18,22): Related location: this is the postconditi
-	this : Patterns.Simple = (p := @0)
-	s : seq<char> = ['A']
-	@0 : seq<char> = ['?']
+assume exists boundVar0: char, boundVar1: char :: this != null && |this.p| > 0 && |s| > 0 && s[0] != '?' && 1 == |s| && 1 == |this.p| && '?' == this.p[0] && s[0] != boundVar0 && s[0] != boundVar1 && boundVar0 != s[0] && boundVar1 != s[0] && boundVar0 != '?' && boundVar1 != '?' && boundVar0 != boundVar1 && boundVar1 != boundVar0;
```
Excellent. There are a number of reasons we'll want to do that, and it's great that it'll be shared.

Dargones commented:
@keyboardDrummer, I made some changes that should make new counterexamples easier to read. Does this look similar to what you are suggesting? In particular, the counterexample for the method in the PR description now looks like this:

```Dafny
assume Node.Cons(Node.Cons(Node.Cons(Node.Nil, 3), 2), 1) == list 
        && Node.Cons(Node.Cons(Node.Cons(Node.Nil, 3), 2), 1).View.requires() 
        && [1, 2, 3] == Node.Cons(Node.Cons(Node.Cons(Node.Nil, 3), 2), 1).View() 
        && Node.Cons(Node.Cons(Node.Nil, 3), 2).View.requires() 
        && [2, 3] == Node.Cons(Node.Cons(Node.Nil, 3), 2).View() 
        && Node.Cons(Node.Nil, 3).View.requires() 
        && [3] == Node.Cons(Node.Nil, 3).View() 
        && Node.Nil.View.requires() 
        && [] == Node.Nil.View();
```

I think this should be easier to read than the previous version! 

Every line after the first one is technically redundant but this is only because the `View` function has a body and a strong postcondition. Were it not so, there would be no way to infer what the function returns from the datatype value itself, so in the general case it makes sense for the counterexample to explicitly constrain all function calls, I think.

keyboardDrummer commented:
> I think this should be easier to read than the previous version!

For sure.

However, I'm still don't understand well why you want to constrain the outputs of functions in the counterexample. Given the pure nature of functions, the outputs of these functions are determined by the input values of the method under test right? It seems like you would always get duplicated information if you assume the input test values and the outputs of functions that depend on these values.

keyboardDrummer commented:
> Second, if `View` ... its specification were not strong enough, we would want the counterexample to constrain the return value. And because one cannot determine how strong a specification is at compile time, we have to constrain return values at all times, including in this case.

I see. If the function has a body and is revealed in the test, doesn't that imply that it is fully specified, and no fresh information is provided by showing its return values?

> In practice, the counterexample will only constrain return values that are specifically mentioned in the Z3 counterexample model, so the number of such constraints is bounded.

Dargones commented:
I think the only situation in which you can drop a function constraint for certain is if (i) the function has a body, and (ii) all function arguments are fully constrained, including the receiver. Both are true for this example but making this check would be non-trivial in general, I think.

keyboardDrummer commented:
It seems like there's two ways of using the unknown-model. One is to generate a test and then run the program, to check whether this model is a counter-example, and the other is to generate an assumption that's placed at the start of the operation under test.

Are you enabling users to use both approaches when asking the Dafny CLI for a counter-example? How does a user get the test generating behavior?

Dargones commented:
Yes, the users can either ask for counterexamples in the form of assertions, or run `dafny generate-tests` to produce test cases!

keyboardDrummer commented:
> Edit: right now, one can only get a counterexample in the form of assumptions. To enable test generation behavior, the counterexample model would need to be cross-referenced with the source code, so this is something that could be left as a future PR.

I think that would be valuable, because if it can be generated, and it does cause an assertion to fail at runtime, it provides a guaranteed counter-example, instead of a counter-example guess (the model produces after an `unknown` result).

Dargones commented:
Have just pushed a commit that adds a warning to notify the user that the counterexample may not be valid due to the solver returning UNKOWN. I would prefer to leave counterexample execution for future work, since it will be a substantial change/addition to the PR. Executing a counterexample is, I think, indeed the only certain way to determine that the counterexample is valid but it is also not clear how applicable such a feature would be (a lot of Dafny code is not intended for execution and code that is intended for execution can fail verification for reasons that would not come up at runtime).

keyboardDrummer commented on docs/DafnyRef/UserGuide.md:
```diff
@@ -2469,12 +2469,12 @@ Legacy options:
+  Note that Danfy cannot guarantee that the counterexample
+  it reports provably violates the assertion or that the assumptions are not
+  mutually inconsistent (see [^smt-encoding]), so this output should be 
+  expected manually and treated as a hint.
```
expected => inspected

keyboardDrummer commented:
> Have just pushed a commit that adds a warning to notify the user that the counterexample may not be valid due to the solver returning UNKOWN. I would prefer to leave counterexample execution for future work, since it will be a substantial change/addition to the PR. Executing a counterexample is, I think, indeed the only certain way to determine that the counterexample is valid but it is also not clear how applicable such a feature would be (a lot of Dafny code is not intended for execution and code that is intended for execution can fail verification for reasons that would not come up at runtime).

Sounds good, thanks. I'm not sure what the best terminology is for these counterexample guesses, but we can iterate on that.

keyboardDrummer commented on docs/DafnyRef/UserGuide.md:
```diff
@@ -2469,12 +2469,12 @@ Legacy options:
-  where to write the counterexample, as well as the
-  `-proverOpt:O:model_compress=false` and
-  `-proverOpt:O:model.completion=true` options.
+* `-extractCounterexample` - if verification fails, report a possible
```
subjective nitpick: possible => potential

Dargones replied:
Done!

keyboardDrummer commented:
@Dargones This PR changes the counterexample API for `dafny server`. Do you know if it would be possible to get back the old API ? I think for the IDE an API that only shows either concrete values or holes, would be ideal.

I imagine an API like;

```c#
record CounterExampleItem(Position Position, IDictionary<string, Value> Variables, IDictionary<int, Value> Heap);
interface Value;
record PrimitiveValue(string content) : Value;
record Unknown() : Value;
record Reference(int number) : Value;
record StructuredValue(string description, IDictionary<string, Value> children) : Value; // description could be the name of the type, or just "object"
```

At some point I think counterexamples in the IDE should be shown using the debugger UI, which would mean we'd have to implement the debug adapter protocol, which has an API not too different from the one above:

https://microsoft.github.io/debug-adapter-protocol/specification#Types_Variable

Dargones commented:
HI, @keyboardDrummer! Here is a quick PR I made that should fix this: https://github.com/dafny-lang/dafny/pull/5847 I agree that in the long term it would be great to use the debugger UI, though this might take one/two weeks of full time work...This PR (as in the one I am leaving the comment on) does a lot of heavy lifting towards that goal, I think.
